### DISCUSSION TITLE
"git pull"  stop working with fatal error

### DISCUSSION THREAD

**Discussion by webquest-nz** [community/community#63768] [Other Feature Feedback, Questions, & Ideas] [answered]
### Select Topic Area

Question

### Body

I have a local codebase and a GitHub repository that have some how diverged according to the error messages I get when trying to push to it.  This apparently means I have commits in each that are not in the other.  Not sure how that happened.  The problem I have is that when I try to pull from it in order to reconcile the extra commit it gives me a fatal error.

`[xx@HP-ProBook-650 racingcp]$ git checkout master
Switched to branch 'master'
Your branch and 'origin/master' have diverged,
and have 2 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
[xx@HP-ProBook-650 racingcp]$ git pull
fatal: Not possible to fast-forward, aborting.
[xx@HP-ProBook-650 racingcp]$ 
`

How should I resolve this?  What I would like to do is to be able to create a new branch and checkout the origin/master head to a local branch so I can reconcile the differences but then I can't even find the syntax to do this.   Is anyone able to help?

**Comment by nirav-gajera** [comment]
The error message "fatal: Not possible to fast-forward, aborting" means that you cannot simply update your local branch to the latest version of the remote branch. This is because your local branch has some commits that are not in the remote branch, and vice versa.

To resolve this, you need to do a merge. There are two ways to do this:

1. **Fast-forward merge:** This is the simplest way to merge, but it is only possible if your local branch is completely contained within the remote branch. In other words, all of the commits in your local branch must also be in the remote branch. If this is the case, you can do a fast-forward merge by running the following command:

```
git pull --ff-only
```

2. **Merge:** This is the more general way to merge, and it can be used even if your local branch is not contained within the remote branch. However, it can also result in merge conflicts, which you will need to resolve manually. To do a merge, run the following command:

```
git pull
```

If you get merge conflicts, you will need to resolve them by editing the conflicting files. Once you have resolved all of the conflicts, you can stage the changes and commit them.

To create a new branch and checkout the origin/master head to a local branch, you can run the following commands:

```
git checkout -b new_branch origin/master
```

This will create a new branch called `new_branch` and checkout the `origin/master` head to it.

Once you have created the new branch, you can merge the changes from the `master` branch into it. You can do this using either the fast-forward merge or the merge method described above.

**Reply by webquest-nz** [reply]
My GitHub Repo looks like its about six months old and I need to make it accept it's changes and reject it's changes.  Looking at the diff it seems as though there are a whole bunch of files that are not in my local copy and yet works on my laptop so it's likely I discarded them some time ago and if not then I can simply recommit them again.  (They look like all 3rd party modules).

```
git checkout -b new_branch origin/master
```
This did actually work.  What do I do if I want to accept the changes in my local repo but reject those from origin/master.  How would I perform that a merge strategy at this point and get github repo back to a useable state again?

**Comment by EnriquezAlberto** [comment]
1_ Create a New Local Branch: Start by creating a new local branch to work on resolving the differences.

`git checkout -b new_branch`

2_ Pull from the Remote Branch: Pull from the remote branch (origin/master) to bring remote changes into your newly created local branch.

`git pull origin master`

3_ Resolve Conflicts if Any: Conflict may arise if changes have been made to the same lines on both ends. Open the conflicting files, manually resolve the differences, save the changes, and then commit.


# After resolving conflicts
```
git add .
git commit -m "Conflict resolution"
```
4_ Finalize the Merge Process: Once conflicts are resolved and committed, you can push your changes to the remote branch.


`git push origin new_branch`

5_ Create a Pull Request if Needed: If you want to merge your changes with the main branch (master) on the remote repository, you can do so through a pull request on GitHub.

1. Go to your GitHub repository.
2. Create a pull request comparing your new branch (new_branch) to the master branch.
3. Provide details about the changes made and complete the pull request process.

By following these steps, you should be able to resolve differences between your local diverged branch and the remote branch, and then merge your changes appropriately. Remember to always back up or make a copy of your code before making significant changes to your repository.

**Reply by webquest-nz** [reply]
When I try step 4 I get:

```
[xxx@HP-ProBook-650 racingcp]$ git checkout diverged
Switched to branch 'diverged'
[xxx@HP-ProBook-650 racingcp]$ git pull origin master
From https://github.com/webquest-nz/racingcp
 * branch            master     -> FETCH_HEAD
fatal: Not possible to fast-forward, aborting.
[xxx@HP-ProBook-650 racingcp]$ 

```

**Comment by webquest-nz** [comment, marked as answer]
```
git merge origin/master
```

seemed to work but with merge conflicts.  The merge conflicts were quite easy to resolve as the changes were obvious.

After that I was able to commit them back to GitHub and  then do a successful git push.

However the error messages from Git were very very misleading.

**Reply by praveshshrestha** [reply]
thanks worked for me as well
